paparazzi/src/main/java/app/cash/paparazzi/internal/apng/APNGReader.kt

```diff
@@ -0,0 +1,265 @@
+    if (!header.contentEquals(PNG_SIG)) {
+      throw IllegalStateException("Input is not a valid PNG")
+    }
+
+    processMetadata()
+  }
+
+  fun getNextFrame(): BufferedImage? {
+    if (finished()) return null
+
+    if (frameNumber == 0) {
+      processTo(IDAT)
+    } else {
+      processTo(FDAT)
+    }
+
+    return getFrame()
+  }
+
+  fun getDelay(): Pair<Int, Int> {
+    return frameInfo?.let { it.delayNumerator.toInt() to it.delayDenominator.toInt() } ?: (1 to 0)
+  }
+
+  fun reset() {
+    frameNumber = 0
+    sequenceNumber = -1
+    frameInfo = null
+
+    fileHandle.reposition(buffer, PNG_SIG.size.toLong())
+    processMetadata()
+  }
+
+  private fun processMetadata() {
+    processTo(IHDR)
+
+    if (buffer.peekNextChunkId().contentEquals(ACTL.bytes)) {
+      processTo(ACTL)
+    }
+  }
+
+  fun finished() = buffer.peekNextChunkId().contentEquals(IEND.bytes)
+
+  override fun close() {
+    fileHandle.close()
+    buffer.close()
+  }
+
+  private fun processTo(header: Header) {
+    do {
+      val (currentHeader, data) = buffer.readChunk()
+      when (currentHeader) {
+        IHDR -> data.parseIHDR()
+        ACTL -> data.parseACTL()
+        FCTL -> data.parseFCTL()
+        IDAT -> data.parseIDAT()
+        FDAT -> data.parseFDAT()
+        IEND -> throw IllegalStateException("PNG ended while processing to $header")
+      }
+    } while (!header.bytes.contentEquals(currentHeader.bytes) && !finished())
+  }
+
+  private fun BufferedSource.parseIHDR() {
+    width = readInt()
+    height = readInt()
+    scratchSpace = IntArray(width * height)
+
+    skip(1L) // Bit depth
+    hasAlpha = when (readByte()) {
+      PNG_COLOR_TYPE_RGB -> false
+      PNG_COLOR_TYPE_RGBA -> true
+      else -> throw IllegalStateException("Unsupported color model")
+    }
+  }
+
+  private fun BufferedSource.parseACTL() {
+    frameCount = readInt()
+  }
+
+  private fun BufferedSource.parseFCTL() {
+    val nextSequence = readInt()
+    frameInfo = FrameInfo(
+      width = readInt(),
+      height = readInt(),
+      offsetX = readInt(),
+      offsetY = readInt(),
+      delayNumerator = readShort(),
+      delayDenominator = readShort()
+    )
+
+    if (sequenceNumber + 1 != nextSequence) {
+      throw IllegalStateException("Out of order sequence, expected: ${sequenceNumber + 1} actual: $nextSequence")
+    }
+    sequenceNumber = nextSequence
+  }
+
+  private fun BufferedSource.parseIDAT() {
+    decompress().decodeToScratchSpace()
+  }
+
+  private fun BufferedSource.parseFDAT() {
+    val nextSequence = readInt()
+    if (sequenceNumber + 1 != nextSequence) {
+      throw IllegalStateException("Out of order sequence, expected: ${sequenceNumber + 1} actual: $nextSequence")
+    }
+    sequenceNumber = nextSequence
+    decompress().decodeToScratchSpace()
+  }
+
+  private fun BufferedSource.decompress(): Buffer {
+    val inflater = Inflater().apply {
+      setInput(readByteArray())
```

Comment:
Future optimization: Use Okio's built-in `InflaterSource` to avoid having to allocate this giant array. The built-in one will share the backing byte arrays that you don't get to see directly with the `Inflater` to avoid copies / allocations.